File: JavaScript/834-Sum of Distances in Tree.js
/**
 * @param {number} n
 * @param {number[][]} edges
 * @return {number[]}
 */
var sumOfDistancesInTree = function(n, edges) {
    let count = new Array(n).fill(0), subRes = new Array(n).fill(0);
    
    let graph = new Map();
    for (let [start, end] of edges){
        graph.set(start, (graph.get(start) || new Array()).concat([end]) );
        graph.set(end, (graph.get(end) || new Array()).concat([start]) );
    }
    
    let dfs = (curr, prev) =>{
        for (let nxt of graph.get(curr)||[] ){
            if (nxt === prev) continue;
            dfs(nxt, curr);
            subRes[curr] += (count[nxt] + subRes[nxt]);
            count[curr] += count[nxt];
        }
        count[curr] += 1;
    }
    
    let dfs2 = (curr, prev) =>{
        for (let nxt of graph.get(curr)||[] ){
            if (nxt === prev) continue;
            subRes[nxt] += (n-count[nxt]) + (subRes[curr] - subRes[nxt] - count[nxt]);
            dfs2(nxt, curr);
        }
    }
    
    dfs(0, null);
    dfs2(0, null);
    return subRes;
};

File: C++/834-Sum of Distances in Tree.cpp
class Solution {
public:
    vector<int> sumOfDistancesInTree(int n, vector<vector<int>>& edges) {
        vector<int> res(n, 0);
        vector<int> subCount(n, 0);
        vector<vector<int>> graph(n);

        for (auto &edge:edges) {
            graph[edge[0]].push_back(edge[1]);
            graph[edge[1]].push_back(edge[0]);
        }

        function<int(int, int)> dfs = [&](int curr, int prev) {   // cannot use auto here
            for (int &nxt : graph[curr]) {
                if (nxt == prev) continue;
                subCount[curr] += dfs(nxt, curr);
                res[curr] += subCount[nxt] + res[nxt];
            }
            subCount[curr]++;
            return subCount[curr];
        };

        function<void(int, int)> dfs2 = [&](int curr, int prev) {   // cannot use auto here
            for (int &nxt:graph[curr]) {
                if (nxt == prev) continue;
                res[nxt] = res[curr] + (n-subCount[nxt]) - subCount[nxt];  // res of parent + (added) - (substracted)
                dfs2(nxt, curr);
            }
        };

        dfs(0, -1);
        dfs2(0, -1);

        return res;
    }
};